docs/2-Interviews/answering_technical_questions.md

```diff
@@ -1,39 +1,129 @@
+- Just to make sure, it's worth asking permission, or at least make sure the interviewer knows what the function is
+- It's fine to make up a function if you can't remember what the existing one is called. Nobody cares if you use `File.open(...)` when it's supposed to be `Sys.open_file(...)`
+
+On that subject, a very useful bit of preparation that can only take an hour or two is to familiarize yourself with the libraries that come with your language. It's worth reading over the operations you can do on strings, arrays/lists and maps/dictionaries. Useful functions include `zip`, `split`, `substr`/`slice`, `join`, `find`, etc.
+
+### Algorithms: finishing the interview
+
+- Once you have an initial implementation mostly right, ask if any further optimizations are necessary
+  - Are you optimizing for time? Space? Which is more important?
+- Offer to make further improvements the code. This tends to impress interviewers
+  - How can the style of your code be improved?
+  - Is there any redudancy that could be eliminated?
+  - Could you refactor your solution to be more generalized? e.g. accept an arbitrary iterable data structure as an input when the question just asked you to handle arrays
+  - Can you write the solution more concisely? e.g. you used a bunch of for loops when you could have used map/filter/reduce
+- Write/talk through some test cases if you have time, to prove to yourself and the interviewer that your code works
+  - Volunteering to do this also impresses interviewers, because it shows that you take responsibility for making sure your code is correct
+
+### Algorithms: advanced tips
+
+- Implement functions in breadth-first order, not depth-first order
+  - For example, supposed you have your main function f1 that calls f2 and f3. Function f2 calls f4. An effective implementation order is f1, f2, f3, f4. Don't implement part of f1, jump to f2, then f3, and end up coming back to f1 later. You'll lose track of what you were doing. Instead, just call the functions, explaining to the interviewer what the function does and promise that you'll implement it after you're done with your current function
+- If the solution to a problem is already familiar to you or you can find the solution instantly, be careful not to code it directly without any explanations
+  - There's a danger that the interviewer will think you just memorized the solution, which doesn't provide them much information about your problem solving-abilities
+  - To mitigate this, explain the constraints of the problems, the conditions that need to be satisfied for the solution to work, and the reasoning behind each part of the solution. For example, if your solution involves using a set, explain that you need constant-time access, that you don't need to store duplicate elements, that you have no values (keys only) and that order doesn't matter (unless the set implementation in your language guarantees an order)
```

Comment:
You should probably also tell the interviewer you know the answer from before, that you have already figured it out. They might give you another question, or you might be asked to still implement it, just explaining the solution as you go